src/Controllers/DatabaseController.js

```diff
@@ -683,20 +688,63 @@ class DatabaseController {
 
   // Adds a relation.
   // Returns a promise that resolves successfully iff the add was successful.
-  addRelation(key: string, fromClassName: string, fromId: string, toId: string) {
+  async addRelation(key: string, fromClassName: string, fromId: string, toId: string) {
+    const className = `_Join:${key}:${fromClassName}`;
     const doc = {
       relatedId: toId,
       owningId: fromId,
     };
+    await this.createJoinTable(className);
     return this.adapter.upsertOneObject(
-      `_Join:${key}:${fromClassName}`,
+      className,
       relationSchema,
       doc,
       doc,
       this._transactionalSession
     );
   }
 
+  createJoinTable(className: string) {
+    const startupPromise = async () => {
+      if (this._relationTablesCache.classes.includes(className)) {
+        return;
+      }
+      const exists = await this.collectionExists(className);
+      const names = [];
+      if (exists) {
+        const indexes = (await this.adapter.getIndexes(className)) || [];
+        names.push(
+          ...indexes.map(({ name, indexname }) => {
+            if (name) {
+              return name.split('_')[0];
+            }
+            const splitName = indexname.split('_');
+            return splitName.at(-1);
+          })
+        );
+      }
+      const keys = ['relatedId', 'owningId'];
+      await Promise.all(
+        keys.map(async subKey => {
+          if (names.includes(subKey)) {
+            return;
+          }
+          try {
+            await this.adapter.ensureIndex(className, relationSchema, [subKey]);
+          } catch (error) {
+            if (error.code !== '23505') {
+              logger.warn('Unable to create relatedId index: ', error);
+            }
+          }
+        })
+      );
+      this._relationTablesCache.classes.push(className);
+    };
+    const promise = this._relationTablesCache.promises[className] || startupPromise();
+    this._relationTablesCache.promises[className] = promise;
+    return promise;
+  }
+
```

Comment:
```suggestion
  ensureJoinTable(className: string) {
    const classNamePromise = this._relationTablesCache[className]
    if(classNamePromise) return classNamePromise
    const startupPromise = async () => {
      const exists = await this.collectionExists(className);
      const names = [];
      if (exists) {
        const indexes = (await this.adapter.getIndexes(className)) || [];
        names.push(
          ...indexes.map(({ name, indexname }) => {
            if (name) {
              return name.split('_')[0];
            }
            const splitName = indexname.split('_');
            return splitName.at(-1);
          })
        );
      }
      const keys = ['relatedId', 'owningId'];
      await Promise.all(
        keys.map(async subKey => {
          if (names.includes(subKey)) {
            return;
          }
          try {
            await this.adapter.ensureIndex(className, relationSchema, [subKey]);
          } catch (error) {
            if (error.code !== '23505') {
              logger.warn('Unable to create relatedId index: ', error);
            }
          }
        })
      );
      this._relationTablesCache.classes.push(className);
    };
    const promise = startupPromise()
    this._relationTablesCache[className] = promise
    return promise
  }

```

Too many source of truth just use a signe object with indexed promises by class names and early return
Code not tested but the strategu should work